File: frontend/src/types/sketch.ts
export type SnapNodeTier = "primary" | "secondary";

export type EnrichedSnapNode = {
  id: string;
  x: number;
  y: number;
  z: number;
  tier: SnapNodeTier;
  elementId: string;
  elementType: string;
  /** 0 = start, 1 = end, or fractional position along member */
  paramAlongMember: number;
};

export type SketchDrawPhase = "idle" | "picking" | "dialogue";

export type StructuralIntentKind =
  | "tie_beam"
  | "bracing"
  | "purlin"
  | "beam"
  | "unknown";

export type SketchApplyScope = "all_bays" | "row" | "single";

export type StructuralIntentResult = {
  kind: StructuralIntentKind;
  confidence: number;
  label: string;
  angleClass: "horizontal" | "vertical" | "diagonal";
  spanMm: number;
  start: { elementType: string; z: number; elementId: string };
  end: { elementType: string; z: number; elementId: string };
};

export type SketchLockedLine = {
  start: EnrichedSnapNode;
  end: EnrichedSnapNode;
};

export type SketchProfileTier = "light" | "recommended" | "conservative";

export type SketchProfileOption = {
  profile: string;
  tier: SketchProfileTier;
  tier_label: string;
  utilization: number;
  governing: string;
};

export type SketchAnalysisResult = {
  intent: StructuralIntentResult;
  profiles: SketchProfileOption[];
  message: string;
  scope_suggestion: SketchApplyScope;
  scope_reason: string;
  alternatives: StructuralIntentKind[];
  ai_available: boolean;
  operations?: import("@/types/structural-advise").OperationProposal[];
  recommended_operation_id?: string | null;
  summary?: string;
};

export type SketchSession = {
  phase: SketchDrawPhase;
  firstNodeId: string | null;
  lockedLine: SketchLockedLine | null;
  intent: StructuralIntentResult | null;
  intentOverride: StructuralIntentKind | null;
  analysis: SketchAnalysisResult | null;
  analysisLoading: boolean;
  selectedOperationId: string | null;
  dialogueStep: 1 | 2 | 3;
  selectedProfile: string | null;
  applyScope: SketchApplyScope | null;
};

export const SKETCH_INTENT_OPTIONS: { kind: StructuralIntentKind; label: string }[] =
  [
    { kind: "tie_beam", label: "Tie Beam" },
    { kind: "bracing", label: "Bracing" },
    { kind: "purlin", label: "Purlin" },
    { kind: "beam", label: "Beam" },
  ];

File: backend/schemas/sketch.py
"""Sketch-mode AI analysis request/response models."""

from typing import Literal

from pydantic import BaseModel, Field

from schemas.elements import ProjectElementMm
from schemas.site import SiteContext

StructuralIntentKind = Literal[
    "tie_beam", "bracing", "purlin", "beam", "unknown"
]
SketchApplyScope = Literal["all_bays", "row", "single"]
AngleClass = Literal["horizontal", "vertical", "diagonal"]
ProfileTier = Literal["light", "recommended", "conservative"]
SnapNodeTier = Literal["primary", "secondary"]


class SketchSnapNode(BaseModel):
    x: float
    y: float
    z: float
    element_id: str
    element_type: str
    tier: SnapNodeTier = "primary"
    param_along_member: float = 0.0


class SketchAnalyzeRequest(BaseModel):
    project_elements: list[ProjectElementMm] = Field(default_factory=list)
    start_node: SketchSnapNode
    end_node: SketchSnapNode
    intent_override: StructuralIntentKind | None = None
    site_context: SiteContext | None = None
    shed_params: dict | None = None
    x_coords_mm: list[float] | None = None
    z_coords_mm: list[float] | None = None


class SketchIntentResult(BaseModel):
    kind: StructuralIntentKind
    confidence: float
    label: str
    angle_class: AngleClass
    span_mm: float
    start_element_type: str
    end_element_type: str
    start_element_id: str
    end_element_id: str


class SketchProfileOption(BaseModel):
    profile: str
    tier: ProfileTier
    tier_label: str
    utilization: float
    governing: str


class SketchAnalyzeResponse(BaseModel):
    intent: SketchIntentResult
    profiles: list[SketchProfileOption]
    message: str
    scope_suggestion: SketchApplyScope
    scope_reason: str
    alternatives: list[StructuralIntentKind] = Field(default_factory=list)
    ai_available: bool = False
